src/functional/otf/compilation/build_systems/compiledb.py

```diff
@@ -0,0 +1,293 @@
+    and not in number of GT4Py programs. In cases where many programs can reuse the same configuration,
+    this can save multiple seconds per program over rerunning CMake configuration each time.
+
+    Works independently of what is used to generate the compiledb.
+    """
+
+    root_path: pathlib.Path
+    source_files: dict[str, str]
+    fencil_name: str
+    compile_commands_cache: pathlib.Path
+    bindings_file_name: str
+
+    def build(self):
+        self.write_files()
+        if build_data.read_data(self.root_path).status < build_data.BuildStatus.CONFIGURED:
+            self.run_config()
+        if (
+            build_data.BuildStatus.CONFIGURED
+            <= build_data.read_data(self.root_path).status
+            < build_data.BuildStatus.COMPILED
+        ):
+            self.run_build()
+
+    def write_files(self):
+        for name, content in self.source_files.items():
+            (self.root_path / name).write_text(content, encoding="utf-8")
+
+        build_data.write_data(
+            data=build_data.BuildData(
+                status=build_data.BuildStatus.STARTED,
+                module=pathlib.Path(""),
+                entry_point_name=self.fencil_name,
+            ),
+            path=self.root_path,
+        )
+
+    def run_build(self):
+        logfile = self.root_path / "log_build.txt"
+        compile_db = json.loads((self.root_path / "compile_commands.json").read_text())
+        assert compile_db
+        with logfile.open(mode="w") as log_file_pointer:
+            for entry in compile_db:
+                log_file_pointer.write(entry["command"] + "\n")
+                subprocess.check_call(
+                    entry["command"],
+                    cwd=self.root_path,
+                    shell=True,
+                    stdout=log_file_pointer,
+                    stderr=log_file_pointer,
+                )
+
+        build_data.update_status(new_status=build_data.BuildStatus.COMPILED, path=self.root_path)
+
+    def run_config(self):
+        compile_db = json.loads(self.compile_commands_cache.read_text())
+
+        (self.root_path / "build").mkdir(exist_ok=True)
+        (self.root_path / "bin").mkdir(exist_ok=True)
+
+        for entry in compile_db:
+            for key, value in entry.items():
+                entry[key] = (
+                    value.replace("$NAME", self.fencil_name)
+                    .replace("$BINDINGS_FILE", self.bindings_file_name)
+                    .replace("$SRC_PATH", str(self.root_path))
+                )
+
+        (self.root_path / "compile_commands.json").write_text(json.dumps(compile_db))
+
+        build_data.write_data(
+            build_data.BuildData(
+                status=build_data.BuildStatus.CONFIGURED,
+                module=pathlib.Path(compile_db[-1]["output"]),
+                entry_point_name=self.fencil_name,
+            ),
+            self.root_path,
+        )
+
+
+def _cc_cache_name(
```

Comment:
The dependencies are already included in the cache system. The build type and the build flags aren't, but they should be. That would mean that this function does not need to exist, because it does not make the cache folder any more unique than it already is.

Could you perhaps add the build type and build flags into the existing cache system, and delete this function?

Reply:
That would include updating the existing cache system to work on `CompilableSource` level, because this function also takes the binding's dependencies into account. In addition it is unclear how to pass build system configuration to the cache system, because the cache system is invoked before the build system. Furthermore this would require changes to allow to build system agnostically find and return the cached program in the compilation step.

This is not to say this work shouldn't be done, only to ask should it be done here?

Reply:
The `Compiler` has access to the `CMakeFactory` at the moment it asks for the cache directory. Build system factories could have a method that returns the hash of the build configuration, and only the hash is passed to the cache system. What do you think? If you don't want to do it now, can you make an issue for this?

Reply:
That should be broadly the way to go but I would prefer to do that when changing the cache system anyway, so I will make an issue

Reply:
https://github.com/GridTools/gt4py/issues/985